hbase-server/src/test/java/org/apache/hadoop/hbase/client/AbstractTestUpdateConfiguration.java

```diff
@@ -0,0 +1,103 @@
+ */
+package org.apache.hadoop.hbase.client;
+
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.nio.file.StandardCopyOption;
+import org.apache.hadoop.hbase.HBaseTestingUtility;
+import org.apache.hadoop.hbase.util.JVMClusterUtil.RegionServerThread;
+
+/**
+ * Base class to test Configuration Update logic. It wraps up things needed to
+ * test configuration change and provides utility methods for test cluster setup,
+ * updating/restoring configuration file.
+ */
+public abstract class AbstractTestUpdateConfiguration {
+  private static final String SERVER_CONFIG = "hbase-site.xml";
+  private static final String OVERWRITE_SERVER_CONFIG  = "overwrite-hbase-site.xml";
+  private static final String BACKUP_SERVER_CONFIG  = "backup-hbase-site.xml";
+
+  private static Path configFileUnderTestDataDir;
+  private static Path overwriteConfigFileUnderTestDataDir;
+  private static Path backupConfigFileUnderTestDataDir;
+
+  protected static void setUpConfigurationFiles(final HBaseTestingUtility testUtil)
+    throws Exception {
+    // Before this change, the test will update hbase-site.xml under target/test-classes and
+    // trigger a config reload. Since target/test-classes/hbase-site.xml is being used by
+    // other testing cases at the same time, this update will break other testing cases so it will
+    // be flakey in nature.
+    // To avoid this, the change is to make target/test-classes/hbase-site.xml immutable. A new
+    // hbase-site.xml will be created under its test data directory, i.e,
+    // hbase-server/target/test-data/UUID, this new file will be added as a resource for the
+    // config, new update will be applied to this new file and only visible to this specific test
+    // case. The target/test-classes/hbase-site.xml will not be changed during the test.
+
+    String absoluteDataPath = testUtil.getDataTestDir().toString();
+
+    // Create test-data directories.
+    Files.createDirectories(Paths.get(absoluteDataPath));
+
+    // Copy hbase-site.xml from target/test-class to target/test-data/UUID directory.
+    Path configFile = Paths.get("target", "test-classes", SERVER_CONFIG);
+    configFileUnderTestDataDir = Paths.get(absoluteDataPath, SERVER_CONFIG);
+    Files.copy(configFile, configFileUnderTestDataDir);
+
+    // Copy overwrite config file overwrite-hbase-site.xml from target/test-class to
+    // target/test-data/UUID directory.
+    Path overwriteConfigFile = Paths.get("target", "test-classes",
+      OVERWRITE_SERVER_CONFIG);
+    overwriteConfigFileUnderTestDataDir = Paths.get(absoluteDataPath, OVERWRITE_SERVER_CONFIG);
+    Files.copy(overwriteConfigFile, overwriteConfigFileUnderTestDataDir);
+
+    backupConfigFileUnderTestDataDir = Paths.get(absoluteDataPath, BACKUP_SERVER_CONFIG);
+
+    // Add the new custom config file to Configuration
+    testUtil.getConfiguration().addResource(testUtil.getDataTestDir(SERVER_CONFIG));
```

Comment:
nit: `testUtil.getDataTestDir(SERVER_CONFIG)` is the same path as `configFileUnderTestDataDir` ? why not reuse that existing variable?

Reply:
Yeah, you are right!

Reply:
Oh No, my bad. configFileUnderTestDataDir is java Path and addResource() needs a hdfs Path, so cannot reuse here.